xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -669,17 +1140,287 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
+      // Skip initial convert, if present.
+      if (instr->user_count() == 1 &&
+          instr->users()[0]->opcode() == HloOpcode::kConvert &&
+          CompatibleElementType(instr->users()[0])) {
+        instr = instr->users()[0];
+      }
+
+      // Verify the uniqueness of the captured Custom Call and inputs.
+      if (!fwd_custom_call.Instr() || !x.Instr() || !dy.Instr() ||
+          !scale.Instr() || !fused_expectation.Instr() ||
+          !fused_norm_factor.Instr()) {
+        VLOG(1) << "Layer norm gradient inputs not unique.";
+        return OkStatus();
+      }
+
+      // Retrieve meta data of the forward layer norm.
+      auto norm_meta_data =
+          norm_meta_data_.find(fwd_custom_call.Instr()->unique_id());
+      if (norm_meta_data == norm_meta_data_.end()) {
+        VLOG(1) << "Unable to retrieve norm meta data of forward Custom Call.";
+        return OkStatus();
+      }
+
+      // Verify the dimensions of reductions in the backward graph.
+      if (reduce0->dimensions() != norm_meta_data->second.norm_dims ||
+          reduce1->dimensions() != norm_meta_data->second.norm_dims ||
+          reduce2->dimensions() != norm_meta_data->second.norm_dims ||
+          reduce3->dimensions() != norm_meta_data->second.norm_dims) {
+        VLOG(1) << "Unexpected reductions dimensions in layer norm gradient.";
+        return OkStatus();
+      }
+
+      // The captured scalar must be one over the number of elements in the
+      // broadcasted dimensions.
+      float actual_r_nelems = scalar->literal().GetAsDouble({}).value();
+      int64_t nelems = 1;
+      for (int i = 0; i < broadcast->shape().dimensions_size(); ++i) {
+        if (!c_linear_search(broadcast->dimensions(), i)) {
+          nelems *= broadcast->shape().dimensions()[i];
+        }
+      }
+      // The absolute of the difference between the actual scaling factor and
+      // the reference value must not exceed a prescribed threshold.
+      float r_nelems = 1. / static_cast<float>(nelems);
+      float numerical_epsilon = std::numeric_limits<bfloat16>::epsilon();
+      if (!(abs(actual_r_nelems - r_nelems) <
+            ((actual_r_nelems + r_nelems) * numerical_epsilon))) {
+        VLOG(1)
+            << "Layer norm backward broadcast operand outside expected range.";
+        return OkStatus();
+      }
+
+      // Identify Dscale = AddReduce(DY * XCenter * norm factor) assuming the
+      // form MultiplyAnyOrder((factor1 or norm factor),
+      // MultiplyAnyOrder(factor0 * (factor 1 or norm factor))).
+      auto find_dscale =
+          [&fused_norm_factor, &norm_meta_data](
+              const UniqueHloInstruction& factor0,
+              const UniqueHloInstruction& factor1) -> HloInstruction* {
+        for (const HloInstruction* factor0_user : factor0.Instr()->users()) {
+          std::vector<HloInstruction*> users;
+          SkipUnaryOpsTopDownRecursive(
+              const_cast<HloInstruction*>(factor0_user), users);
```

Comment:
Instead of a const_cast, can SkipUnaryOpsTopDownRecursive operate on const pointers?